### ISSUE TITLE
Can not consume events from other COM classes via WithEvents

### ISSUE THREAD

**Issue by tdimitriou** [WaynePhillipsEA/twinbasic#83]
**Describe the bug**
The application will compile, but is not able to run, when I declare an object (coming from external referenced COM Library) WithEvents, 

**To Reproduce**
This is the code to reproduce the problem. The project references Olaf Schmidt's RichClient 6 (RC6). I have also tested with other COM Dlls (created with VB6) with the same results. The program will run just fine if I do not declare the Form WithEvents inside the MainWindow Class.

```
Module modMain
    Public New_c as RC6.cConstructor = New RC6.cConstructor
    
    Sub Main()
        
        Dim Window As MainWindow = New MainWindow

        Window.Show
        
        New_c.Cairo.WidgetForms.EnterMessageLoop
                
    End Sub
    
End Module

Class MainWindow
    Public WithEvents Form As RC6.cWidgetForm
    
    Sub New()
        Set Form = New_c.Cairo.WidgetForms.Create()
    End Sub
    
    Public Sub Show(Optional ByVal Modal as FormShowConstants = vbModeless)
        Form.Show Modal
    End Sub 
    
    Private Sub Form_Load()
        Form.Caption = "Hello, World!"
    End Sub    
End Class

```

**Desktop (please complete the following information):**
 - OS: Windows 10
 - twinBASIC compiler version: 0.9.1882

**Additional context**
The output of the debug console is as follows:
```
[LINKER] SUCCESS created output file 'c:\Users\Doros\Downloads\twinBASIC_RC6\Build\RC6Project_win32.exe'
[RUN] process 11024 started.
[RUN] process 11024 exited with code -1073741819

```

**Comment by WaynePhillipsEA** [comment]
Thanks for reporting.  I am aware of this.  For WithEvents to work, it currently depends on whether the source event interface is early-bound (i.e. vtable bound) or late bound (i.e. IDispatch- bound).    Most event sinks are defined late-bound, and so require the class implementations to support IDispatch.  Right now, twinBASIC does not offer IDIspatch support.  In fact, it is what I am working on right now 😁

IDispatch support will be included in a release early next week, and then this shall be resolved.

**Comment by WaynePhillipsEA** [comment]
Should be fixed in v0.9.2021, as twinBASIC now offers IDispatch support for twinBASIC classes.

**Comment by vbRichClient** [comment]
Can confirm, that Event-support seems to work as it should (at least in a small Demo with the RC6 Form-Engine).

![TwinCairoClock](https://user-images.githubusercontent.com/4271823/116316624-84dc9d00-a7b2-11eb-9856-8905db6ce442.png)

The Demo-Code is less than 100 lines, so I'm putting it here:
```
Option Explicit

Module modHelloRC6
  Public New_c As New cConstructor, Cairo As cCAiro 'the two RC6-MainConstructor-Objects
  Public fMain As New cfMain, UDummy As cUTC 'our MainForm-Class + UDummy, to avoid a tB-Bug

  Sub Main()
    Set Cairo = New_c.Cairo 'derive a global cairo-instance from New_c
        Cairo.SetDPIAwareness '<- we use this here, to avoid a manifest-resource
        Cairo.ImageList.AddImage "bgForm",  wget("http://vbRichClient.com/Downloads/bgSquares.jpg")
        Cairo.ImageList.AddImage "bgClock", wget("http://vbRichClient.com/Downloads/clockFace.png")
    
    fMain.Form.Show 'after loading 2 ImageResources from the web, we now show the MainForm
    
    Cairo.WidgetForms.EnterMessageLoop 'and finally enter the MessagePump (we exit from it, when Cairo.WidgetForms.Count reaches 0)
  End Sub
  
  Public Function wget(URL As String) As Byte()
    Dim http As Object = CreateObject("WinHttp.WinHttpRequest.5.1") 
        http.Open "GET", URL, 0: http.Send
    return http.ResponseBody
  End Function
End Module

Class cfMain
  Public  WithEvents Form As cWidgetForm 
  Private VLayout As cLayoutVertical, HBand as cLayoutBand
  Private WithEvents grpTzOpt as cWidgetBase, imgClock As cwImage
    
  Private Sub Class_Initialize()
    Set Form = Cairo.WidgetForms.Create(, "Hello RC6-Forms", , 1024, 800)
        Form.WidgetRoot.ImageKey = "bgForm"
        Form.WidgetRoot.Zoom = Form.WidgetRoot.CurrentMonitor.Zoom 'ensure DPI-awareness
        Form.IconImageKey = "bgClock"
        Form.SetMinMaxDimensions 800, 700
        Form.TimerInterval = 100 '<- VBA-UserForm-compatible Timer-Prop 
        
    Set VLayout = New cLayoutVertical
        VLayout.AttachTo Form
        
    Dim CWs(0 To 1): CWs(0) = 0.3: CWs(1) = 0.7 '(CellWidths > 1 are treated as "absolute", values <=1 relative in percent)
    Set HBand = VLayout.AddBandWithCells(CWs) 'Add a horizontal band with two cells to the layout
    'Set HBand = VLayout.AddBandWithCells(0.3, 0.7) '<- use this line instead of the 2 above, as soon as TB supports ParamArrays

    Set grpTzOpt = HBand.Cell(0).BeginGroupVertical("") 'a vertical group-frame (for a few Option-Boxes)
        HBand.Cell(0).AddOptionBox "Local", True '<- pre-select this Option-Button
        HBand.Cell(0).AddOptionBox "Cairo" 'yeah - pun intended... ;-)
        HBand.Cell(0).AddOptionBox "Tokyo"
        HBand.Cell(0).AddOptionBox "Mumbai"
        HBand.Cell(0).AddOptionBox "Honolulu"
    HBand.Cell(0).EndGroup 'tells the VLayout-cell, that a group is complete (BTW, .BeginGroupHorizontal is available too)
    grpTzOpt.Alpha = 0.55 '<- just to demonstrate the "nested Alpha-Rendering-support" of the Widget-Engine
    
    'the second cell in this horizontal band will take up only our ImgBox-Widget
    Set imgClock = HBand.Cell(1).AddImageBox("", "bgClock", 501, 501)  '<- define BG-ImageKey and Canvas-Size
  End Sub

  Private Sub grpTzOpt_BubblingEvent(Sender As Object, EventName As String, P1, P2, P3, P4, P5, P6, P7) 'OptBtn-Group-Handler
    If EventName = "Click" Then Form.Tag = Sender.Caption 'simply store the current Caption
  End Sub

  Private Sub Form_Timer()
    Dim UTC As cUTC = New_c.UTC, T As Date = UTC.SystemUTC
    Select Case Form.Tag '<- we've stored the last clicked OptionBtn-Caption there
      Case "Local":    T += UTC.LocalOffs '<- SystemUTC + LocalOffs is equivalent to T = Now
      Case "Cairo":    T += UTC.OffsFromStr("+02:00")
      Case "Tokyo":    T += UTC.OffsFromStr("+09:00")
      Case "Mumbai":   T += UTC.OffsFromStr("+05:30")
      Case "Honolulu": T += UTC.OffsFromStr("-10:00")
    End Select
    'note, that the (non-Local)) TimeZones above, are ones without any DayLight-Saving-Offsets
    '(one will have to determine and add those Extra-DST-Offsets for TimeZones where DayLight-Saving applies)
    
    If imgClock.Widget.Tag <> T Then imgClock.Widget.Tag = T Else Exit Sub 'avoid Canvas-refreshs, if there was no change on T

    imgClock.Cls 'clear any Drawings (which were made on top of the existing ClockFace-BG-Image)
      DrawClockHand imgClock.Canvas, 110, 8, &H80, 30 * (Hour(T) Mod 12 + Minute(T) / 60)
      DrawClockHand imgClock.Canvas, 160, 4, &H808000, 6 * (Minute(T) + Second(T) / 60)
      DrawClockHand imgClock.Canvas, 220, 1, &H8000, 6 * (Second(T))
      DrawCenterDot imgClock.Canvas, Cairo.CreateRadialPattern(0, 0, 10, -3.8, -3.8, 0)
    imgClock.Refresh

    Form.Caption = Form.Tag & "-Clock  " & T
  End Sub
  
  '**** Cairo-Drawing-Helpers - the CC is the typical Cairo-Context (in C-examples usually named cr)
  Private Sub DrawClockHand(CC As cCairoContext, Length, LineWidth, ByVal Color&, AngleDeg)
    CC.SaveAndTransform CC.Surface.Width / 2, CC.Surface.Height / 2, AngleDeg 
      CC.DrawLine 0, Length * -0.8, 0, Length * 0.2, False, LineWidth, Color
    CC.Restore
  End Sub

  Private Sub DrawCenterDot(CC As cCairoContext, Pat As cCairoPattern)
    Pat.AddGaussianStops_TwoColors vbBlack, vbWhite, 1, 0.9
    CC.SaveAndTransform CC.Surface.Width / 2, CC.Surface.Height / 2 
      CC.Arc 0, 0, 10
      CC.Fill 0, Pat
    CC.Restore
  End Sub
End Class
```
For those who want to run the code, before you start:
- the newest RC6-Version 6.0.8 needs to be downloaded
- unzipped into a Folder of your choice
- and then RC6.dll and RC6Widgets.dll need to be registered (e.g. via the 2 little install-scripts in the Package)

Then (after registering) - the twinBasic-IDE can be set to include these two references:
![RC6_Referencing](https://user-images.githubusercontent.com/4271823/116315703-2ebb2a00-a7b1-11eb-8de2-176ed349d58d.png)

@WaynePhillipsEA 
In case you read this - there's two (little) problems which I had to workaround:
1) For some weird reason, I needed to define: UDummy As cUTC globally 
... (although an instance of cUTC is used only in a differently named local Variable inside the Form_Timer EventHandler.

2) ParamArrays (when defined on the Interface of an external COM-lib) are not recognized as:
... "open-ended Param-Lists" (worked around that, by defining a VariantArray Dim CWs(0 To 1) - then passing that).

HTH

Olaf

**Comment by WaynePhillipsEA** [comment]
@vbRichClient   Thanks... that's really good to see working!!  I've fixed the issues you raised above in v0.9.2100.   That's going on twitter 👍

**Comment by WaynePhillipsEA** [comment]
Closing this now, any further edge cases can be raised in new issues.  Thanks all.
